endpoints/s3/src/abortmultipartupload.cpp

```diff
@@ -0,0 +1,183 @@
+		path = bucket.value();
+		path = irods::s3::finish_path(path, url.segments());
+		logging::debug("{}: AbortMultipartUpload path={}", __func__, path.string());
+	}
+	else {
+		logging::error("{}: Failed to resolve bucket", __func__);
+		response.result(beast::http::status::forbidden);
+		logging::debug("{}: returned {}", __func__, response.reason());
+		session_ptr->send(std::move(response));
+		return;
+	}
+
+	// get the uploadId from the param list
+	std::string upload_id;
+	if (const auto upload_id_param = url.params().find("uploadId"); upload_id_param != url.params().end()) {
+		upload_id = (*upload_id_param).value;
+	}
+
+	if (upload_id.empty()) {
+		logging::error("{}: Did not receive a an uploadId", __func__);
+		response.result(beast::http::status::bad_request);
+		logging::debug("{}: returned {}", __func__, response.reason());
+		session_ptr->send(std::move(response));
+		return;
+	}
+
+	// Do not allow an upload_id that is not in the format we have defined. People could do bad things
+	// if we didn't enforce this.
+	if (!std::regex_match(upload_id, upload_id_pattern)) {
+		logging::error("{}: Upload ID {} was not in expected format.", __func__, upload_id);
+		response.result(beast::http::status::bad_request);
+		logging::debug("{}: returned {}", __func__, response.reason());
+		session_ptr->send(std::move(response));
+		return;
+	}
+
+	// delete the entry in the replica_token_and_number_map
+	if (part_shmem::replica_token_and_number_map.find(upload_id) == part_shmem::replica_token_and_number_map.end()) {
+		// Read all of the shared pointers in the tuple to make sure they are destructed in
+		// the order we require. std::tuple does not guarantee order of destruction.
+		auto conn_ptr = std::get<2>(part_shmem::replica_token_and_number_map[upload_id]);
+		auto transport_ptr = std::get<3>(part_shmem::replica_token_and_number_map[upload_id]);
+		auto dstream_ptr = std::get<4>(part_shmem::replica_token_and_number_map[upload_id]);
+		if (dstream_ptr) {
+			dstream_ptr->close();
+		}
+
+		// delete the entry
+		part_shmem::replica_token_and_number_map.erase(upload_id);
+	}
+
+	// get the base location for the part files
+	const nlohmann::json& config = irods::http::globals::configuration();
+	std::string part_file_location =
+		config.value(nlohmann::json::json_pointer{"/s3_server/multipart_upload_part_files_directory"}, ".");
+
+	// remove the temporary part files - <part_file_location>/irods_s3_api_<upload_id>.[N]
+	std::string part_file_regex_str = "irods_s3_api_" + upload_id + "\\.[0-9]+";
+	std::regex part_files_regex(part_file_regex_str);
+	const std::filesystem::directory_iterator end;
+	try {
+		for (std::filesystem::directory_iterator iter{part_file_location}; iter != end; ++iter) {
+			if (std::filesystem::is_regular_file(*iter)) {
+				if (std::regex_match(iter->path().filename().string(), part_files_regex)) {
+					std::filesystem::remove(iter->path());
+				}
+			}
+		}
+	}
+	catch (std::exception&) {
+	}
```

Comment:
I reorganized the code to log an error message and set the return code to internal_server_error.  Rather than the errors above I allow it to continue so that the shmem will be cleaned up.

I also separated the errors between filesystem iterator type exceptions and file removal exceptions so better error messages can be provided.